src/app/core/services/notifications/message-handler.service.ts

```diff
@@ -0,0 +1,57 @@
+import { Injectable } from '@angular/core'
+import { FirebaseX } from '@ionic-native/firebase-x/ngx'
+
+import { UsageEventType } from '../../../shared/enums/events'
+import { Assessment, AssessmentType } from '../../../shared/models/assessment'
+import { MessagingAction } from '../../../shared/models/notification-handler'
+import { QuestionnaireService } from '../config/questionnaire.service'
+import { LogService } from '../misc/log.service'
+import { ScheduleService } from '../schedule/schedule.service'
+import { UsageService } from '../usage/usage.service'
+
+@Injectable()
+export class MessageHandlerService {
+  constructor(
+    public firebase: FirebaseX,
+    public logger: LogService,
+    public schedule: ScheduleService,
+    public questionnaire: QuestionnaireService,
+    public usage: UsageService
+  ) {
+    this.firebase
+      .onMessageReceived()
+      .subscribe(data => this.onMessageReceived(new Map(Object.entries(data))))
+  }
+
+  onMessageReceived(data) {
+    const action = data.get('action')
+    switch (action) {
+      case MessagingAction.QUESTIONNAIRE_TRIGGERED:
+        this.logger.log('A questionnaire was triggered!')
+        const questionnaire = <Assessment>JSON.parse(data.get('questionnaire'))
+        const metadata = data.get('metadata')
+        return this.triggerQuestionnaire(questionnaire).then(() =>
+          this.usage.sendQuestionnaireEvent(
+            UsageEventType.QUESTIONNAIRE_TRIGGERED,
+            questionnaire.name,
+            Date.now(),
+            metadata
```

Comment:
if this is a map type in avro will it automatically be converted to map from json?

Reply:
Made the changes to convert the json to map.